source/common/upstream/upstream_impl.cc

```diff
@@ -262,13 +262,48 @@ ResourceManager& ClusterInfoImpl::resourceManager(ResourcePriority priority) con
   return *resource_managers_.managers_[enumToInt(priority)];
 }
 
-void ClusterImplBase::blockHcUpdates(bool block) {
-  ASSERT(block_hc_updates_ == !block);
-  block_hc_updates_ = block;
+void ClusterImplBase::setInitializeCallback(std::function<void()> callback) {
+  ASSERT(!initialization_started_);
+  ASSERT(initialization_complete_callback_ == nullptr);
+  initialization_complete_callback_ = callback;
+}
+
+void ClusterImplBase::startInitialization() {
+  if (!initialization_started_ && health_checker_ && pending_initialize_health_checks_ == 0) {
+    pending_initialize_health_checks_ = hosts().size();
+
+    // TODO(mattklein123): Remove this callback when done.
+    health_checker_->addHostCheckCompleteCb([this](HostSharedPtr, bool) -> void {
+      if (pending_initialize_health_checks_ > 0 && --pending_initialize_health_checks_ == 0) {
+        finishInitialization();
+      }
+    });
+  }
+
+  if (!initialization_started_ && pending_initialize_health_checks_ == 0) {
+    initialization_started_ = true;
+    finishInitialization();
+  } else {
+    initialization_started_ = true;
```

Comment:
`initialization_started_` is set true on both these branches?

Reply:
this is in response to a previous @alyssawilk comment. I will add a comment.